### PR TITLE
Reworking the rest of the modules

### PR DESCRIPTION
**Pull request by ardizzone**
- [x] all-in-one block
- [x] coupling blocks
- [x] up-/downsampling
- [x] fixed operations
- [x] i-ResNet
- [x] invertible autoencoder
- [x] orthogonal
- [x] gaussian mixture
- [x] graph topology (@wapu ?)

### CODE REVIEW

**Comment by ardizzone** [comment]
Except for the coupling blocks and reshapes, I only changed the call signatures and return values. Hopefully, all should be working now, just so people can go ahead with other tasks.

Code improvements and docstrings are still missing for many (everything except `all_in_one_block.py`, `coupling.py`, and `reshapes.py`, those are completely done I think.

**Comment by ardizzone** [comment]
Ok, feel free to merge, then we can go through and improve the code in detail at our leisure.

File: FrEIA/modules/invertible_resnet.py [hunk 1]
```diff
@@ -134,20 +133,22 @@ def lipschitz_correction(self):
                     self.layers[i].weight.data *= self.spectral_norm_max / spectral_norm
 
 
-    def forward(self, x, c=[], rev=False):
+    def forward(self, x, c=[], rev=False, jac=True):
+        jac = self._jacobian(x, c, rev=rev)
```
Comment (wapu) [inline FrEIA/modules/invertible_resnet.py:137]: This is costly, so put it a check for `jac == True`?

File: tests/reversible_graph_net.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+        self.inp_size = (3, 10, 10)
+        self.cond_size = (1, 10, 10)
+
+        inp = InputNode(*self.inp_size, name='input')
+        cond = ConditionNode(*self.cond_size, name='cond')
+
+        split = Node(inp,
+                     Split,
+                     {'split_size_or_sections': [1,2], 'dim': 0},
```
Comment (wapu) [inline tests/reversible_graph_net.py:40]: Parameter is called `section_sizes` now :)

**Review by wapu — COMMENTED** [review]
I left two comments where I saw a problem, otherwise looks great!

File: FrEIA/modules/invertible_resnet.py (hunk 1 above)
Comment (ardizzone) [reply FrEIA/modules/invertible_resnet.py:137]: Good point, I will do that! :)

File: tests/reversible_graph_net.py (hunk 2 above)
Comment (ardizzone) [reply tests/reversible_graph_net.py:40]: Ah, I think that's someone else's department, I don't know why it's in my PR.
But I will change it nonetheless! :smile:

File: FrEIA/modules/gaussian_mixture.py [hunk 3]
```diff
@@ -146,35 +146,19 @@ def forward(self, x, c, rev=False):
...
             if fixed_components:
                 # Return latent codes of x according to chosen component distributions only
-                return [torch.stack([torch.matmul(U[b,i[b],:,:], x[0][b,:] - mu[b,i[b],:]) for b in range(batch_size)])]
+                return [torch.stack([torch.matmul(U[b,i[b],:,:], x[0][b,:] - mu[b,i[b],:]) for b in range(batch_size)])], self.jac
             else:
                 # Return latent codes of x according to all component distributions simultaneously
                 if len(x[0].shape) < 3:
                     x[0] = x[0][:,None,:]
-                return [torch.matmul(U, (x[0] - mu)[...,None])[...,0]]
+                return [torch.matmul(U, (x[0] - mu)[...,None])[...,0]], self.jac
```
Comment (fdraxler) [inline FrEIA/modules/gaussian_mixture.py:154]: Why self.jac? Jacobians should never be stored

File: FrEIA/modules/inv_auto_layers.py [hunk 4]
```diff
@@ -22,30 +23,30 @@ def log_e(self, s):
         '''log of the nonlinear function e'''
         return self.clamp * 0.636 * torch.atan(s/self.clamp)
 
-    def forward(self, x, rev=False):
+    def forward(self, x, rev=False, jac=True):
+        j = (-1)**rev * torch.sum(self.log_e(self.alpha_pos + 0.5 * (self.alpha_neg - self.alpha_pos) * (1 - x[0].sign())), dim=1)
```
Comment (fdraxler) [inline FrEIA/modules/inv_auto_layers.py:27]: -1 ** rev^^
I liked the if rev: return …, -j else: return …, j

File: FrEIA/modules/inv_auto_layers.py [hunk 5]
```diff
@@ -22,30 +23,30 @@ def log_e(self, s):
...
 
-    def __init__(self, dims_in):
-        super().__init__()
+    def __init__(self, dims_in, dims_c=None):
+        super().__init__(dims_in, dims_c)
         self.alpha = nn.Parameter(0.01 * torch.randn(dims_in[0][0]) + 0.7)
 
-    def forward(self, x, rev=False):
+    def forward(self, x, rev=False, jac=True):
+        # TODO this one didn't even HAVE a jacobian implemented ?!?!
```
Comment (fdraxler) [inline FrEIA/modules/inv_auto_layers.py:45]:
```suggestion
        if jac is True:
            raise NotImplementedError("Jacobian is not implemented for InvAutoAct")
```

File: FrEIA/modules/inv_auto_layers.py [hunk 6]
```diff
@@ -54,48 +55,46 @@ def output_dims(self, input_dims):
...
+class LearnedElementwiseScaling(InvertibleModule):
 
-    def __init__(self, dims_in):
-        super().__init__()
+    def __init__(self, dims_in, dims_c=None):
+        super().__init__(dims_in, dims_c)
         self.s = nn.Parameter(torch.zeros(*dims_in[0]))
 
-    def forward(self, x, rev=False):
+    def forward(self, x, rev=False, jac=True):
```
Comment (fdraxler) [inline FrEIA/modules/inv_auto_layers.py:83]:
```suggestion
    def forward(self, x, rev=False, jac=True):
        j = self.s
```

File: FrEIA/modules/inv_auto_layers.py [hunk 7]
```diff
@@ -54,48 +55,46 @@ def output_dims(self, input_dims):
...
         if not rev:
-            return [x[0] * self.s.exp()]
+            return [x[0] * self.s.exp()], None
```
Comment (fdraxler) [inline FrEIA/modules/inv_auto_layers.py:85]:
```suggestion
            return [x[0] * self.s.exp()], j
```

File: FrEIA/modules/inv_auto_layers.py [hunk 8]
```diff
@@ -54,48 +55,46 @@ def output_dims(self, input_dims):
...
         else:
-            return [x[0] * self.s.neg().exp_()]
+            return [x[0] * self.s.neg().exp_()], None
```
Comment (fdraxler) [inline FrEIA/modules/inv_auto_layers.py:87]:
```suggestion
            return [x[0] * self.s.neg().exp_()], -j
```

File: FrEIA/modules/inv_auto_layers.py [hunk 9]
```diff
@@ -105,21 +104,21 @@ def __init__(self, dims_in, dims_out=None):
         self.weights = nn.Parameter(0.01 * torch.randn(self.dims_out[0][0], self.dims_in[0][0]))
         self.bias = nn.Parameter(0.01 * torch.randn(1, self.dims_out[0][0]))
 
-    def forward(self, x, rev=False):
+    def forward(self, x, rev=False, jac=True):
```
Comment (fdraxler) [inline FrEIA/modules/inv_auto_layers.py:107]:
```suggestion
    def forward(self, x, rev=False, jac=True):
        if jac is True:
            raise NotImplementedError("Jacobian is not implemented for InvAutoFC")
```

File: FrEIA/modules/inv_auto_layers.py [hunk 10]
```diff
@@ -128,15 +127,15 @@ def __init__(self, dims_in, dims_out, kernel_size=3, padding=1):
         self.conv2d = nn.Conv2d(dims_in[0][0], dims_out[0][0], kernel_size=kernel_size, padding=padding, bias=False)
         self.bias = nn.Parameter(0.01 * torch.randn(1, dims_out[0][0], 1, 1))
 
-    def forward(self, x, rev=False):
+    def forward(self, x, rev=False, jac=True):
```
Comment (fdraxler) [inline FrEIA/modules/inv_auto_layers.py:130]:
```suggestion
    def forward(self, x, rev=False, jac=True):
        if jac is True:
            raise NotImplementedError("Jacobian is not implemented for InvAutoConv2D")
```

File: FrEIA/modules/gaussian_mixture.py (hunk 3 above)
Comment (ardizzone) [reply FrEIA/modules/gaussian_mixture.py:154]: in the cases when it's constant, it makes sense to store it!

Comment (fdraxler) [reply FrEIA/modules/gaussian_mixture.py:154]: Agreed.
Then we should have some code that says `if self.jac is not None: jac = ...`, but I only see `self.jac = ...`

**Review by fdraxler — COMMENTED** [review]
Looks good :)

**Pull request merged**